pkg/cgroups/manager/manager.go

```diff
@@ -0,0 +1,392 @@
+				"load-balancing tracing feature will not work")
+			m.enabled = false
+		}
+	})
+
+	if m.enabled {
+		log.Info("Cgroup metadata manager is enabled")
+	}
+}
+
+func (m *CgroupManager) processPodEvents() {
+	for {
+		select {
+		case ev := <-m.podEvents:
+			if !m.enabled {
+				continue
+			}
+			switch ev.eventType {
+			case podAddEvent, podUpdateEvent:
+				m.updatePodMetadata(ev.pod, ev.oldPod)
+			case podDeleteEvent:
+				m.deletePodMetadata(ev.pod)
+			case podGetMetadataEvent:
+				m.getParentPodMetadata(ev.cgroupId, ev.podMetadataOut)
+			}
+		case <-m.shutdown:
+			return
+		}
+	}
+}
+
+func (m *CgroupManager) updatePodMetadata(pod, oldPod *v1.Pod) {
+	id := string(pod.ObjectMeta.UID)
+	pm, ok := m.podMetadataById[id]
+	if !ok {
+		// Fill in pod static metadata.
+		pm = &podMetadata{
+			name:      pod.Name,
+			namespace: pod.Namespace,
+		}
+		m.podMetadataById[id] = pm
+	}
+	if oldPod != nil && oldPod.Status.DeepEqual(&pod.Status) {
+		return
+	}
+	// Only update the metadata that can change. This excludes pod's name,
+	// namespace, id, and qos class.
+	podIPs := pod.Status.PodIPs
+	pm.ips = make([]string, len(podIPs))
+	for i := range podIPs {
+		pm.ips[i] = podIPs[i].IP
+	}
+	// Get metadata for pod's containers that are in the running state. Containers
+	// can get re-created, and their ids can change. Update the new containers.
+	// Pod's metadata including its containers map will be deleted when the pod
+	// is deleted.
+	numContainers := len(pod.Status.ContainerStatuses)
+	if pm.containers == nil && numContainers > 0 {
+		pm.containers = make(map[string]struct{})
+	}
+	currContainers := make(map[string]struct{}, numContainers)
+	for _, c := range pod.Status.ContainerStatuses {
+		var cId string
+		if cId = c.ContainerID; cId == "" || c.State.Running == nil {
+			continue
+		}
+		// The container ID field is of the form: <container-runtime>://<containerID>
+		// Example:containerd://e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0e
+		_, after, found := strings.Cut(cId, "//")
+		if !found || after == "" {
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+			}).Errorf("unexpected container ID: %s", cId)
+			continue
+		}
+		cId = after
+		if _, ok := pm.containers[cId]; ok {
+			currContainers[cId] = struct{}{}
+			// Container cgroup path doesn't change as long as the container id
+			// is the same.
+			continue
+		}
+		pm.containers[cId] = struct{}{}
+		currContainers[cId] = struct{}{}
+
+		// Container could've been gone, so don't log any errors.
+		cgrpPath, err := m.getContainerCgroupPath(id, cId, pod.Status.QOSClass)
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+			}).Debugf("failed to get container metadata for (%s): %v", cId, err)
+			continue
+		}
+		cgrpId, err := m.cgroupsChecker.GetCgroupID(cgrpPath)
```

Comment:
If we ever get restricted on the underlying syscalls/fs access by something like SELinux here, I think we're going to wonder why the feature silently doesn't work :thinking: 

Not sure what the best option is to handle this. I agree we don't want to be printing errors all the time for expected conditions, but also if something slightly unusual is going on then this is going to break silently and it's going to be difficult to debug. Hmm.

Reply:
Hmm... Not sure what the best way to handle this either. Perhaps use the cilium pod as a reference, and log a one time error?

Reply:
:bulb: ah interesting, yeah so just do a one-off detection on startup that should be a sign of whether this call would work, and emit an error in that case since that should be enough to indicate whether there is likely ever a problem with this or not. :+1:

Reply:
Cilium pod might not be the very first pod that the cgroup manager receives events for. I've added a TODO in the code to revisit this as there are unknowns such as what would the errors even look like.